docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+differ in security. Therefore, developers are strongly encouraged to use
+encryption for comprehensive protection.
+
+`oe_unseal()` decrypts a sealed blob **in-place**. The code snippet below shows
+how to decrypt a sealed blob and retrieve a pointer to the plaintext. Please
+note that `my_data` after this call will point to somewhere within `blob`.
+
+```C
+result = oe_unseal(blob, blob_size,
+                   &my_data,        // plaintext
+                   &my_data_size,   // plaintext_size
+                   NULL,            // additional_data
+                   NULL             // additional_data_size
+                   );
+if (result != OE_OK)
+    // Handle decryption error...
+else
+{
+    // Process my_data...
+}
+```
+
+The initial `key_info` passed to `oe_seal()` could be retrieved by a call to
+`oe_get_seal_key_by_policy()`, which takes a `seal_policy` as its only input
+parameter. In practice, an enclave may desire to encrypt different things under
+different keys. Per
+[NIST.SP800-108](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf),
+*Label* and *Context* should be among the inputs to *KDF* (*Key Derivation
+Function*). Most TEEs also support taking addtional *context* or *entropy* into
+key derivation. The existing `oe_get_seal_key_by_policy()` API however lacks
+that support and prompts for a new API to fill the gap.
+
+The new `oe_get_seal_key_info()` API is introduced with the following
+signature.
+
+```C
+oe_result_t oe_get_seal_key_info(
+    oe_seal_policy_t seal_policy,
+    uint8_t* entropy,
+    size_t entropy_size,
+    uint64_t tee_specific,
+    uint8_t** key_info,
+    size_t* key_info_size);
+```
+
+If compared to `oe_get_seal_key_by_policy()`, `oe_get_seal_key_info()` differs
+in:
+
+- The addition of `entropy`/`entropy_size` parameters - This allows *Label* or
+  *Context* (or both), or simply more entropy to be mixed into the seal key.
+- The addition of `tee_specific` parameter - This is a TEE specific parameter,
+  or `OE_SEAL_SGX` to use TEE defaults. See
+  [oe_get_seal_key_info()](#oe_get_seal_key_info()) later in this document for
+  more information.
+- The removal of `key_buffer`/`key_size` output parameters - Direct use of seal
+  keys is error prone so discouraged.
+
+The code snippet below shows how to create a `key_info` suitable for a
+provisioning enclave. It adds `PROVISION_KEY` to the attribute mask, and uses
+`PROVISION_SEAL_KEY` to exclude CPU SVN from key derivation.
+
+```C
+sgx_key_request_t *key_info;
+size_t key_info_size;
+result = oe_get_seal_key_info(OE_SEAL_POLICY_PRODUCT, NULL, 0,
+                              OE_SEAL_SGX | OE_SEALKEY_DEFAULT_FLAGSMASK |
+                              SGX_FLAGS_PROVISION_KEY,
+                              (uint8_t**)&key_info, &key_info_size);
+if (result != OE_OK)
+    // Handle errors...
+else
+    key_info->key_name = SGX_KEYSELECT_PROVISION_SEAL;
```

Comment:
What does this line do?   Why is it necessary to do this on the result rather than passing sufficient inputs to the function in 202?   This looks cumbersome.

Reply:
`key_name` is a very SGX specific feature and is needed only for provisioning enclaves. 

Generally speaking, each TEE has its distinct features, so it's difficult to define a generic API that works for all TEE-specific features. And it'd be overly complicated even if we could manage to define one. Thus I'm shooting for a "reasonable" API definition that would work for the majority (like 90%+) without blocking the minority. E.g., this corner case of provisioning enclave wouldn't be blocked, even though an extra step would be required.

Reply:
I'd suggest using the same model as we use for oe_create_enclave().    Allow a TEE-specific buffer to be optionally passed in.

Reply:
I did consider the TEE-specific buffer model but I decided to do away from that, because it's hard to specify some structure fields and use TEE default for others. E.g., for `KEYREQUEST` on SGX, how to specify a custom `ATTRIBUTEMASK` while using defaults for all other fields? The typical approach, which is also what I'm following here, is to get a structure with all defaults filled in, and modify its fields as needed.